src/game/client/worldheightmap.cpp

```diff
@@ -29,21 +45,2358 @@ class GDIFileStream : public InputStream
+    } else {
+        m_borderSize = file.Read_Int32();
+    }
+
+    if (info->version < 4) {
+        m_boundaries.resize(1);
+        m_boundaries[0].x = m_width - 2 * m_borderSize;
+        m_boundaries[0].y = m_height - 2 * m_borderSize;
+    } else {
+        int size = file.Read_Int32();
+        m_boundaries.resize(size);
+
+        for (int i = 0; i < size; i++) {
+            m_boundaries[i].x = file.Read_Int32();
+            m_boundaries[i].y = file.Read_Int32();
+        }
+    }
+
+    m_dataSize = file.Read_Int32();
+    m_data = new unsigned char[m_dataSize];
+
+    if (m_dataSize <= 0 || m_dataSize != m_height * m_width) {
+        throw CODE_05;
+    }
+
+    m_siesmicUpdateWidth = (m_width + 7) / 8;
+    m_seismicUpdateFlags = new unsigned char[m_height * m_siesmicUpdateWidth];
+    Clear_Seismic_Update_Flags();
+    m_siesmicZVelocity = new float[m_dataSize];
+    Fill_Seismic_Z_Velocities(0.0f);
+    file.Read_Byte_Array(m_data, m_dataSize);
+
+    if (info->version == 1) {
+        int x = (m_width + 1) / 2;
+        int y = (m_height + 1) / 2;
+
+        for (int i = 0; i < y; i++) {
+            for (int j = 0; j < x; j++) {
+                m_data[j + y * i] = m_data[2 * j + m_width * 2 * i];
+            }
+        }
+
+        m_width = x;
+    }
+
+    return true;
+}
+
+bool WorldHeightMap::Parse_Blend_Tile_Data_Chunk(DataChunkInput &file, DataChunkInfo *info, void *user_data)
+{
+    return ((WorldHeightMap *)user_data)->Parse_Blend_Tile_Data(file, info, user_data);
+}
+
+bool WorldHeightMap::Parse_Blend_Tile_Data(DataChunkInput &file, DataChunkInfo *info, void *user_data)
+{
+    int size = file.Read_Int32();
+
+    if (m_dataSize != size) {
+        throw CODE_05;
+    }
+
+    m_tileNdxes = new short[m_dataSize];
+    m_cliffInfoNdxes = new short[m_dataSize];
+    m_blendTileNdxes = new short[m_dataSize];
+    m_extraBlendTileNdxes = new short[m_dataSize];
+    m_flipStateWidth = (m_width + 7) / 8;
+    m_cellFlipState = new unsigned char[m_height * m_flipStateWidth];
+    m_cliffState = new unsigned char[m_height * m_flipStateWidth];
+    memset(m_cellFlipState, 0, m_height * m_flipStateWidth);
+    memset(m_cliffState, 0, m_height * m_flipStateWidth);
+
+    file.Read_Byte_Array((unsigned char *)m_tileNdxes, 2 * m_dataSize);
+    file.Read_Byte_Array((unsigned char *)m_blendTileNdxes, 2 * m_dataSize);
+
+    if (info->version >= 6) {
+        file.Read_Byte_Array((unsigned char *)m_extraBlendTileNdxes, 2 * m_dataSize);
+
+        if (g_theWriteableGlobalData->m_use3WayTerrainBlends) {
+            memset(m_extraBlendTileNdxes, 0, 2 * m_dataSize);
+        }
+    }
+
+    if (info->version >= 5) {
+        file.Read_Byte_Array((unsigned char *)m_cliffInfoNdxes, 2 * m_dataSize);
+    }
+
+    if (info->version < 7) {
+        Init_Cliff_Flags_From_Heights();
+    } else if (info->version == 7) {
+        int w = (m_width + 1) / 8;
+        unsigned char *c = new unsigned char[w * m_height];

```

Comment:
this memory is leaked?

Reply:
yeah looks like it, fix and mark as bugfix

Reply:
fixed.